mpact_api/mpact/views.py

```diff
@@ -83,70 +83,25 @@ def post(self, request):
             }
         }
         """
-        message = request.data.get(constants.MESSAGE)
-        logger.debug(message)
-        chat_data = message.get(constants.CHAT)
-        from_data = message.get(constants.FROM)
+        try:
+            message = request.data.get(constants.MESSAGE)
+            event = TelegramEvents(message)
+            # TELEGRAM_EVENTS is a set of event names :: {
+            # "new_chat_participant",
+            # "new_chat_title",
+            # "text",
+            # }
+            # Here taking out the common event between set and incoming request_data
+            # and calling the method from TELEGRAM_EVENTS class
+            getattr(event, list(constants.TELEGRAM_EVENTS.intersection(message))[0])()
```

Comment:
I appreciate that this is only one line of code, but it needs seven lines of comments to explain what it's doing, and it's a lot less readable than the `if`/`elif` clause it is replacing. If you use `if`/`elif` it's still eight lines, and you don't need the comment.
```suggestion
            if constants.NEW_CHAT_PARTICIPANT in message:
                event.new_chat_participant()
            elif constants.NEW_CHAT_TITLE in message:
                event.new_chat_title()
            elif constants.TEXT in message:
                event.text()
            else:
                raise ValueError(f'Unrecognized event in message: {message!r}')
```

Reply:
Is it ok to expose the message data, if there are some errors? If it is ok, I will add in serializer errors also. It can be used by the admin for manual entry or for some other analytics. Can I get your advice on this?

Reply:
I think we definitely want to raise serializer errors. But you make a very good point. It might be a good idea to anonymize messages. Maybe that last line should say, 

```python
                raise ValueError(f'Unrecognized event in message: {anonymize(message)!r}')
```

where `anonymize()` is something like this:

```python
ANON = '*' * 16
REPLACE = {
    'id': 0,
    'username': ANON,
    'first_name': ANON,
    'last_name': ANON,
    'language_code': '**'
}


def anonymize(dict_: dict) -> None:
    """
    Recursively anonymize values in ``dict_`` in place.

    >>> message = {
    ...     "message_id": 51,
    ...     "from": {
    ...         "id": 1229986355,
    ...         "is_bot": False,
    ...         "first_name": "John",
    ...         "last_name": "Walter",
    ...         "language_code": "en"
    ...     },
    ...     "chat": {
    ...         "id": -433354696,
    ...         "title": "Testing Group",
    ...         "type": "group",
    ...         "all_members_are_administrators": True
    ...     },
    ...     "date": 1604317469,
    ...     "new_chat_members": [
    ...         {
    ...             "id": 1321604047,
    ...             "is_bot": True,
    ...             "first_name": "demo_bot",
    ...             "username": "test_bot"
    ...         }
    ...     ]
    ... }
    >>> anonymize(message)
    >>> message == {
    ...     "message_id": 51,
    ...     "from": {
    ...         "id": 0,
    ...         "is_bot": False,
    ...         "first_name": "****************",
    ...         "last_name": "****************",
    ...         "language_code": "**"
    ...     },
    ...     "chat": {
    ...         "id": 0,
    ...         "title": "Testing Group",
    ...         "type": "group",
    ...         "all_members_are_administrators": True
    ...     },
    ...     "date": 1604317469,
    ...     "new_chat_members": [
    ...         {
    ...             "id": 0,
    ...             "is_bot": True,
    ...             "first_name": "****************",
    ...             "username": "****************"
    ...         }
    ...     ]
    ... }
    True

    """
    for key, value in dict_.items():
        if isinstance(value, dict):
            anonymize(value)
        elif (isinstance(value, list)
              and all(isinstance(v, dict) for v in value)):
            for v in value:
                anonymize(v)
        elif key in REPLACE:
            dict_[key] = REPLACE[key]

```

Reply:
This is good, hiding sensitive information. 
Actually what I wanted to say about the serializer is now we are already raising it but we will only get like this 

"rest_framework.exceptions.ValidationError: {'id': [ErrorDetail(string='chat data with this id already exists.', code='unique')]}"
 
So here if we see this error log, we will not be able to find it was for which id, etc. So here If we will include the whole message data with anonymize() it may be helpful for admin for manual checking.